jackson/src/main/java/feign/jackson/JacksonIteratorDecoder.java

```diff
@@ -0,0 +1,164 @@
+ * Returned iterator auto-close the {@code Response} when it reached json array end or failed to parse stream.
+ * If this iterator is not fetched till the end, it has to be casted to {@code Closeable} and explicity {@code Closeable#close} by the consumer.
+ * <p>
+ * <p>
+ * <p>Example: <br>
+ * <pre><code>
+ * Feign.builder()
+ *   .decoder(JacksonIteratorDecoder.create())
+ *   .doNotCloseAfterDecode() // Required to fetch the iterator after the response is processed, need to be close
+ *   .target(GitHub.class, "https://api.github.com");
+ * interface GitHub {
+ *  {@literal @}RequestLine("GET /repos/{owner}/{repo}/contributors")
+ *   Iterator<Contributor> contributors(@Param("owner") String owner, @Param("repo") String repo);
+ * }</code></pre>
+ */
+public final class JacksonIteratorDecoder implements Decoder {
+
+  private final ObjectMapper mapper;
+
+  JacksonIteratorDecoder(ObjectMapper mapper) {
+    this.mapper = mapper;
+  }
+
+  @Override
+  public Object decode(Response response, Type type) throws IOException {
+    if (response.status() == 404) return Util.emptyValueOf(type);
+    if (response.body() == null) return null;
+    Reader reader = response.body().asReader();
+    if (!reader.markSupported()) {
+      reader = new BufferedReader(reader, 1);
+    }
+    try {
+      // Read the first byte to see if we have any data
+      reader.mark(1);
+      if (reader.read() == -1) {
+        return null; // Eagerly returning null avoids "No content to map due to end-of-input"
+      }
+      reader.reset();
+      return new JacksonIterator<Object>(actualIteratorTypeArgument(type), mapper, response, reader);
+    } catch (RuntimeJsonMappingException e) {
+      if (e.getCause() != null && e.getCause() instanceof IOException) {
+        throw IOException.class.cast(e.getCause());
+      }
+      throw e;
+    }
+  }
+
+  private static Type actualIteratorTypeArgument(Type type) {
+    if (!(type instanceof ParameterizedType)) {
+      throw new IllegalArgumentException("Not supported type " + type.toString());
+    }
+    ParameterizedType parameterizedType = (ParameterizedType) type;
+    if (!Iterator.class.equals(parameterizedType.getRawType())) {
+      throw new IllegalArgumentException("Not an iterator type " + parameterizedType.getRawType().toString());
+    }
+    return ((ParameterizedType) type).getActualTypeArguments()[0];
+  }
+
+  public static JacksonIteratorDecoder create() {
+    return create(Collections.<Module>emptyList());
+  }
+
+  public static JacksonIteratorDecoder create(Iterable<Module> modules) {
+    return new JacksonIteratorDecoder(new ObjectMapper()
+        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+        .registerModules(modules));
+  }
+
+  public static JacksonIteratorDecoder create(ObjectMapper objectMapper) {
+    return new JacksonIteratorDecoder(objectMapper);
+  }
+
+  static final class JacksonIterator<T> implements Iterator<T>, Closeable {
```

Comment:
I think we should make this class `public` and we should suggest people to use `JacksonIterator` (which I would also rename to `CloseableIterator`, optional)

Anyway, on the javadoc we should recommend using the `JacksonIterator`, compiler will warn you about closing it.

In the end, I think the `Iterator` should be dealt with similar to `Stream`  https://github.com/OpenFeign/feign/pull/651/files#diff-ee67453098d0b7bbe061ae7c7826ea96R74